### ISSUE TITLE
bug: communication breaks down after ~15 minutes of usage

### ISSUE THREAD

**Issue by ruabmbua** [neojou/rtw88-usb#35]
The wifi still appears to be connected, but no packets make it through the interface.

System:

```
CPU: 8-Core AMD Ryzen 7 3800X (-MT MCP-) speed/min/max: 1976/2200/3900 MHz 
Kernel: 5.6.13-arch1-1 x86_64 Up: 1h 42m Mem: 4267.2/32096.7 MiB (13.3%) 
Storage: 931.51 GiB (43.4% used) Procs: 352 Shell: zsh 5.8 inxi: 3.1.00 
Wifi dongle: rtl8822bu -> AC1200 from PIAEK
```

I also tested in qemu-kvm with usb pass-through (for debugging), but no issues there...

If I can help with anything, please tell me. It would be really nice to finally have a proper (upstreamed) driver for this dongle. I am currently using a custom driver (88x2bu), which is a mess (but it works without downtime).

**Comment by ruabmbua** [comment]
Also: The driver does not work at all, if the dongle is used in ehci (usb2) mode.

**Comment by neojou** [comment]
@ruabmbua Thanks for your using. It would be appreciated if you can provide the kernel log. We can just type like "$ dmesg > rtw88-kernel.log" to get the kernel log file, rtw88-kernel.log". Thanks.

**Comment by ruabmbua** [comment]
Ok here is the log. Looks like I can reproduce it by causing lots of traffic.

https://gist.github.com/ruabmbua/00c9914f61bbe9f6784cd8c665f34fe2

**Comment by neojou** [comment]
Thanks a lot for your kind help. We have some findings and solutions for rcu_note_context_switch, will update it next week.

**Comment by neojou** [comment]
Would you mind helping to see if the commit
https://github.com/neojou/rtw88-usb/commit/1c7d28b29d8d2fcfa8d83f285d588d1b069fb58c
can help, thanks.

**Comment by ruabmbua** [comment]
Yes, I will try to test it tomorrow afternoon, thanks for the work!

**Comment by ruabmbua** [comment]
Tested it, seems to work now, although it looks like the link speed maxes out at ~20mbit/s.

I can reach >100mbit/s with the realtek driver.

**Comment by neojou** [comment]
After checking, a better way is to change from ieee80211_iterate_active_interfaces_atomic to ieee80211_iterate_active_interfaces, so I modified again, and there are two commits for this. 

https://github.com/neojou/rtw88-usb/commit/c42be3b8dcde84b30c61744845242376a3a75117
https://github.com/neojou/rtw88-usb/commit/d6a4bff3b76519769b29ff3f72abae3de436921c

FYI. Thanks.

**Comment by ruabmbua** [comment]
Seems like perf switches between 20mbits/s and >80mbits/s from time to time...

dmesg:

[  526.601466] rtwifi_usb 5-2.1:1.0: timed out to flush queue 3
[  526.601471] rtw_usb_deep_ps ===>
[  526.601471] rtw_usb_deep_ps ===>
[  526.601480] rtw_usb_deep_ps ===>
[  526.734808] rtwifi_usb 5-2.1:1.0: timed out to flush queue 3
[  526.734814] rtw_usb_deep_ps ===>
[  526.874809] rtw_usb_deep_ps ===>
[  526.903599] rtw_usb_deep_ps ===>
[  527.104804] rtw_usb_deep_ps ===>

**Comment by neojou** [comment]
Thanks for this. I think there might be some issues not fixed, and it causes the throughput dropped. Is it possible to have some more information before
  [ 526.601466] rtwifi_usb 5-2.1:1.0: timed out to flush queue 3
There might be some clues there. 

Another might be relative to power saving. let me clarify it further.

**Comment by ruabmbua** [comment]
https://gist.github.com/ruabmbua/36d115c7a4f27c1efec2308c4a15e9af

Yeah looks like powersaving is way to aggressive. It spams rtw_usb_deep_ps really fast, and it is 35mbit/s slow. Takes a while until the rtw_usb_deep_ps gets less frequent, and performance goes up to about 80mbit/s

**Comment by neojou** [comment]
Thanks a lot for this. The log helps. 
There seems another rcu_note_context_switch problem occurred at ieee80211_iface_work
Please let me see how to fix it well ^^ 

  145.973115]  rtw_usb_write8+0x62/0x80 [rtwusb]
[  145.973121]  rtw_bf_enable_bfee_su+0x64/0xe0 [rtw88]
[  145.973127]  rtw_bf_assoc+0x1dd/0x220 [rtw88]
[  145.973132]  rtw_ops_bss_info_changed+0x1c5/0x200 [rtw88]
[  145.973148]  drv_bss_info_changed+0x7e/0x1a0 [mac80211]
[  145.973168]  ieee80211_assoc_success.constprop.0+0x9d7/0xb41 [mac80211]
[  145.973188]  ieee80211_sta_rx_queued_mgmt.cold+0x322/0x459 [mac80211]
[  145.973213]  ieee80211_iface_work+0x1f4/0x350 [mac80211]

**Comment by ruabmbua** [comment]
@neojou Was anything significantly changed, which might resolve the bug? I`d be happy to try the driver again, and make a few notes on what works, and what breaks.

**Comment by neojou** [comment]
@ruabmbua Really appreciate for your kind help. Yes, the current code at the master should be able to solve the problem. Also we found some better approach, and now I'm in test. Plan to release to the master for trial next week.
